packages/gateway/src/structures/Shard.ts

```diff
@@ -0,0 +1,441 @@
+      let buffer = Buffer.alloc(0);
+      buffer = Buffer.concat([buffer, data as Buffer]);
+      const zlibSuffix = Buffer.from([0x00, 0x00, 0xff, 0xff]);
+      if (
+        buffer.length >= 4 &&
+        Buffer.compare(zlibSuffix, buffer.subarray(buffer.length - 4)) === 0
+      )
+        this.inflate?.push(buffer, Z_SYNC_FLUSH);
+      else this.inflate?.push(buffer, false);
+      data = Buffer.from(this.inflate?.result as string);
+    }
+    data = JSON.parse(data.toString());
+    this.handleOPCodes(data);
+  }
+
+  /**
+   * Handles the 'close' event received from the gateway.
+   * @param code The close code.
+   * @param reason The reason for closing.
+   */
+  private _onClose(code: number, reason: RawData) {
+    this.status = 'DISCONNECTED';
+    this.emit('shardClose', this.id, code, reason);
+
+    switch (code) {
+      case GatewayCloseCodes.UnknownError:
+      case GatewayCloseCodes.UnknownOpcode:
+      case GatewayCloseCodes.DecodeError:
+      case GatewayCloseCodes.NotAuthenticated:
+      case GatewayCloseCodes.InvalidSeq:
+      case GatewayCloseCodes.RateLimited:
+      case GatewayCloseCodes.SessionTimedOut:
+        this.reconnect();
+        break;
+      case GatewayCloseCodes.AuthenticationFailed:
+      case GatewayCloseCodes.InvalidShard:
+      case GatewayCloseCodes.ShardingRequired:
+      case GatewayCloseCodes.InvalidAPIVersion:
+      case GatewayCloseCodes.InvalidIntents:
+      case GatewayCloseCodes.DisallowedIntents:
+        break;
+      default:
+        this.reconnect();
+        break;
+    }
+  }
+
+  /**
+   * Handles the 'error' event received from the gateway.
+   * @param error The error received from the gateway.
+   */
+  private _onError(error: ErrorEvent) {
+    this.emit('shardError', this.id, error);
+  }
+
+  /**
+   * Disconnects from the gateway.
+   */
+  disconnect() {
+    if (this.ws?.readyState === WebSocket.OPEN) {
+      this.ws.close();
+      this.status = 'DISCONNECTED';
+      this.emit('shardDisconnect', this.id);
+    }
+  }
+
+  /**
+   * Reconnects to the gateway.
+   */
+  private reconnect() {
+    if (this.status !== 'DISCONNECTED') return;
+    this.status = 'RECONNECTING';
+    if (this.heartbeatInterval) clearInterval(this.heartbeatInterval);
+    this.disconnect();
+
+    let attempts = this.manager.gateway.reconnectAttempts ?? Infinity;
+    this.reconnectInterval = setInterval(() => {
+      if (!attempts) {
+        if (this.reconnectInterval) clearInterval(this.reconnectInterval);
+        this.status = 'DISCONNECTED';
+        return;
+      }
+      this.initWS();
+      attempts--;
+    }, 5000);
+  }
+
+  /**
+   * Sends an identify payload to the gateway.
+   */
+  private async identify() {
+    if (this.ws?.readyState === WebSocket.OPEN) {
+      await sleep(
+        1000 *
+          this.id *
+          (5 /
+            (this.manager.gateway._gatewayData?.session_start_limit
+              .max_concurrency ?? 1))
+      );
```

Comment:
This should be moved into the ShardingManager. Shard shouldn't need to worry about this.
This also wouldn't work if we only wanted to spawn shards 16-32, for example.

Reply:
Moved to ShardingManager.